### PR TITLE
Total Turret Refactor

### PR DESCRIPTION
**Pull request by MarkSuckerberg**
## About The Pull Request
Turret code was absolute garbage. This goes through and re-does practically all of it, plus adds some new goodies for good measure.

Adds my own half-baked "simple beam" system, for when performance is allegedly important and you don't care that they look ugly and only show up when you can see the source object.

## Why It's Good For The Game
Not spending so much time processing JUST turrets is super nice. Plus giving players the ability to actually control their turrets more is nice.

## Changelog

:cl:
add: You can now set turrets to filter by faction, mob type, and a few more criteria including "dangerous only" as well as the ability to disable retaliation.
add: Turret retaliation is a lot more vengeful. Don't mess with them.
add: Turrets will now spend a short time targetting you, pointing a beam at you to indicate that they're doing so.
del: Turret covers, because they sucked.
fix: Turretcode is a LOT less laggy.
tweak: Turrets are now built like normal machines. The boards are currently not available (except through salvaging).
tweak: You can now access full turret settings from the turret control panel.
/:cl:

### CODE REVIEW

File: code/game/machinery/porta_turret/portable_turret.dm [hunk 1]
```diff
@@ -141,247 +166,231 @@ DEFINE_BITFIELD(turret_flags, list(
...
+			to_chat(user, "<span class='warning'>[src] is already in good condition!</span>")
+			return
 
-	else if((I.tool_behaviour == TOOL_WRENCH) && (!on))
-		if(raised)
+		if(!I.tool_start_check(user, amount=0))
 			return
 
+		to_chat(user, "<span class='notice'>You begin repairing [src]...</span>")
+		if(I.use_tool(src, user, 40, volume=50))
```
Comment (Erikafox) [inline code/game/machinery/porta_turret/portable_turret.dm]: what if this was a looping do_after like mecha repairs

File: code/game/machinery/porta_turret/portable_turret_types.dm [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+	lethal_projectile = /obj/projectile/beam/laser/heavylaser
+	lethal_projectile_sound = 'sound/weapons/lasercannonfire.ogg'
+	desc = "An energy blaster auto-turret."
+
+/obj/machinery/porta_turret/syndicate/energy/raven
+	stun_projectile =  /obj/projectile/beam/laser
+	stun_projectile_sound = 'sound/weapons/laser.ogg'
+	faction = list("neutral","silicon","turret")
+
+/obj/machinery/porta_turret/syndicate/pod
```
Comment (Erikafox) [inline code/game/machinery/porta_turret/portable_turret_types.dm:45]: iirc this is used on the assault pod from nukies and can be cut

File: code/game/machinery/porta_turret/portable_turret_types.dm [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+	faction = list(FACTION_PLAYER_INTEQ, "turret")
+
+/obj/machinery/porta_turret/ship/inteq/light
+	name = "Close-In Vanguard Turret"
+	desc = "A light turret designed by IRMG engineers for the the task of defending from close-in encounters. Low power, high speed."
+	stun_projectile = /obj/projectile/bullet/c10mm/rubber
+	stun_projectile_sound = 'sound/weapons/gun/smg/vector_fire.ogg'
+	lethal_projectile = /obj/projectile/bullet/c10mm
+	lethal_projectile_sound = 'sound/weapons/gun/smg/vector_fire.ogg'
+	subsystem_type = /datum/controller/subsystem/processing/fastprocess //turns out if you have a shot delay below what SSmachines fires at you need to use a different subsystem
```
Comment (Erikafox) [inline code/game/machinery/porta_turret/portable_turret_types.dm]: you can probably take this out of fastprocess

File: code/game/machinery/porta_turret/portable_turret_types.dm (hunk 2 above)
Comment (MarkSuckerberg) [reply code/game/machinery/porta_turret/portable_turret_types.dm:45]: nope, it's used on a few maps teehee

File: code/game/machinery/porta_turret/portable_turret.dm (hunk 1 above)
Comment (MarkSuckerberg) [reply code/game/machinery/porta_turret/portable_turret.dm]: smart

File: code/game/machinery/porta_turret/portable_turret_types.dm (hunk 2 above)
Comment (Erikafox) [reply code/game/machinery/porta_turret/portable_turret_types.dm:45]: buh????

**Comment by Erikafox** [comment]
alt-click lock and unlock on turrets could use some feedback (bubble alert or to_chat)

**Comment by MemeSnorfer** [comment]
https://github.com/user-attachments/assets/bcc01520-b44e-46f2-a711-a006735837c3

I'm not sure what's going on here but I don't think this is intended.

**Comment by Erikafox** [comment]
retaliate firing needs to check for if the turret is nonfunctional

**Comment by Erikafox** [comment]
image
Non functional turrets do not clear their lasers

**Comment by MarkSuckerberg** [comment]
oh right yeah you told me this teehee

**Comment by MarkSuckerberg** [comment]
> Shiptest_MokoshZeta2024-11-2716-22-04.mp4
> 
> I'm not sure what's going on here but I don't think this is intended.

next time this happens yell at me to get online so I can check it, genuinely no idea. everything else should be fixed though

**Comment by thgvr** [comment]
lgtm

**Pull request merged**